google/cloud/bigtable/internal/async_retry_unary_rpc_and_poll.h

```diff
@@ -195,6 +195,124 @@ class AsyncRetryAndPollUnaryRpc
+ * A functor used as a continuation to poll for result of a longrunning
+ * operation.
+ *
+ * A simple lambda cannot be used instead because it can't capture the polling
+ * policy, because it's a std::unique_ptr<>, hence, non-copyable.
+ */
+template <typename Client, typename Response>
+class PollAfterRetryFunctor {
+ public:
+  PollAfterRetryFunctor(char const* location,
+                        std::unique_ptr<PollingPolicy> polling_policy,
+                        MetadataUpdatePolicy metadata_update_policy,
+                        std::shared_ptr<Client> client, CompletionQueue cq)
+      : location_(location),
+        polling_policy_(std::move(polling_policy)),
+        metadata_update_policy_(std::move(metadata_update_policy)),
+        client_(std::move(client)),
+        cq_(std::move(cq)) {}
+
+  future<StatusOr<Response>> operator()(
+      future<StatusOr<google::longrunning::Operation>> fut) {
+    auto res = fut.get();
+    if (!res) {
+      return make_ready_future<StatusOr<Response>>(res.status());
+    }
+    return StartAsyncLongrunningOp<Client, Response>(
+        location_, std::move(polling_policy_),
+        std::move(metadata_update_policy_), std::move(client_), std::move(cq_),
+        *std::move(res));
+  }
+
+ private:
+  char const* location_;
+  std::unique_ptr<PollingPolicy> polling_policy_;
+  MetadataUpdatePolicy metadata_update_policy_;
+  std::shared_ptr<Client> client_;
+  CompletionQueue cq_;
+};
+
+/**
+ * Asynchronously start a longrunning operation (with retries) and poll its
+ * result.
+ *
+ * @param location typically the name of the function that created this
+ *     asynchronous retry loop.
+ * @param polling_policy controls how often the server is queried.
+ * @param rpc_retry_policy controls the number of retries, and what errors are
+ *     considered retryable.
+ * @param rpc_backoff_policy determines the wait time between retries.
+ * @param idempotent_policy determines if a request is retryable.
+ * @param metadata_update_policy controls how to update the metadata fields in
+ *     the request.
+ * @param client the client on which `AsyncGetOperation` is called to query
+ *     the longrunning operation's status.
+ * @param async_call the callable to start a new asynchronous operation.
+ * @param request the parameters of the request.
+ * @param cq the completion queue where the retry loop is executed.
+ *
+ * @return a future that becomes satisfied when either the retried RPC or
+ *     polling for the longrunning operation's results fail despite retries or
+ *     after both the initial RPC and the polling for the result of the
+ *     longrunning operation it initiated complete successfully.
+ */
+template <
+    typename AsyncCallType, typename RequestType, typename IdempotencyPolicy,
+    typename Client, typename Response,
+    typename Sig = internal::AsyncCallResponseType<AsyncCallType, RequestType>,
+    typename std::enable_if<Sig::value, int>::type validate_async_call_type = 0,
```

Comment:
In #2236 @devjgm convinced me to not use these, and just rely on SFINAE, you can remove them and save some code here.

Reply:
I'm fine with the code as it is now without the enablers, but I do want to clarify something that I may have been unclear about in #2236. 

I don't think it's a goal to remove all uses of `enable_if`. They do sometimes add value and functionality that's not otherwise there. However, I do think we should avoid enablers when they're redundantly doing something that SFINAE was already doing.

At the sake of over clarifying for future readers, let me state some assertions from which I'll base my conclusion:

*  [SFINAE](https://en.cppreference.com/w/cpp/language/sfinae) only applies to function templates. Not class templates. Not struct meta functions. Just function templates.

* SFINAE only applies to the function signature (or more correctly the *function type*, since it includes the return value). This is sometimes a confusing point. What this means is that a "substitution failure" in a function template *argument* (or return value or template parameter) can result in that overload being removed from the overload set (and thus it's "not an error"). However, a substitution failure within *the body* of the function is not SFINAE -- that's just a normal compile error.

* `std::enable_if<>` gives us the ability to move a non-sfinae compilation error *from the body* into the function signature instead, thus converting a true compilation error into SFINAE, which is "not an error" and instead removes the function from the overload set. `enable_if` is also useful in a `static_assert` within the body of a function, but *this is not sfinae* and instead will simply be a compilation error.

So, I am not opposed to enable_if. It's very useful sometimes. But there's little value in using `enable_if` when the function was *already* relying on sfinae. 

--

Now, in this particular function, there is no other sfinae checks that assure that the AsyncCallType returns a longrunning operation. The enable_ifs were making that signature check a sfinae error. Using a static_assert in the body is fine, and it will produce a pretty clear error message, but it is not sfinae. 

In this particular case, we probably don't care about sfinae vs a clear compile error, so I'm totally OK with a static_assert. Indeed, it's probably a slightly better error message. But I did want to clarify that in #2236 I was not arguing for the removal of all enable_ifs.